### PR TITLE
Added multiple policies for renaming nodes

### PR DESCRIPTION
**Pull request by pierresouchay**
Addressing https://github.com/hashicorp/consul/issues/4414 is hard as many unit tests do rely on having empty nodeIDs.

It seems there are several usages of nodeId:
 - legacy ones (complicated, see https://github.com/hashicorp/consul/pull/4399 )
 - cloud based (see https://github.com/hashicorp/consul/issues/4741 ) where people want to replace a node with the same name quickly
 - bare-metal ones (like us) with people implementing constant nodeIds accross installations.

This PR allows to address those 3 use-cases by adding a new options `node_renaming_policy=(legacy|dead|strict)`

 * `legacy` (the default) will keep the existing behavior
 * `dead`: do not allow nodes without IDs, but allow to replace a dead node with another one with the same name, but different ID (this is DONE by checking that serfHealth is critical). Stilll allow renaming nodes
 * `strict`: do not allow empty nodeID, allow to rename nodes with the same ID, do now allow to steal a node Name as soon as the node is within the catalog.

### CODE REVIEW

**Comment by pierresouchay** [comment]
@mkeeler @banks what do you think of this proposal? (if you remember our conversations in #3983 and #4399, it is a hot topic)

I think it is a good alternative to the current status quo, and will allow to remove progressively the legacy and hard to debug method of current node registration. It will also help people using Consul in Cloud environments as described in #4741 while allowing people like us with bare-metal nodes and fixed nodeids to have clearer policies regarding nodes renaming.

Kind regards

**Comment by banks** [comment]
Scheduling this for attention soon - this issue/proposal is on our radar and will hopefully have cycles to consider it soon.

File: agent/consul/state/catalog.go [hunk 1]
```diff
@@ -354,59 +354,124 @@ func (s *Store) EnsureNode(idx uint64, node *structs.Node) error {
...
+		// There is a dup node, lets check if the node is healthy
+		dupNodeCheck, err := tx.First("checks", "id", dupNode.Node, string("serfHealth"))
+		if err != nil {
+			return nil, fmt.Errorf("Cannot get status of node %s due to: %s", dupNode.Node, err)
+		}
+		if dupNodeCheck == nil {
+			return nil, fmt.Errorf("Cannot validate whether node %s is healthy, cannot rename node", dupNode.Node)
+		}
+		existingDupNodeSerf := dupNodeCheck.(*structs.HealthCheck)
+		if existingDupNodeSerf.Status != "critical" {
```
Comment (Aestek) [inline agent/consul/state/catalog.go]: Use `api.HealthCritical`

File: agent/consul/state/catalog.go [hunk 2]
```diff
@@ -436,6 +527,12 @@ func (s *Store) ensureNodeTxn(tx *memdb.Txn, idx uint64, node *structs.Node) err
 			return nil
 		}
 		node.ModifyIndex = idx
+		if n.Node != node.Node {
```
Comment (Aestek) [inline agent/consul/state/catalog.go]: We are comparing pointers here, shouldn't this be a `.Equal()` ?

File: types/noderenaming.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package types
+
+// Contains definition of policies to rename nodes
+
+const (
```
[the remark this answers is not in the document]
Comment (pierresouchay) [reply types/noderenaming.go]: Ok, DONE in next patch

File: agent/consul/state/catalog.go (hunk 2 above)
Comment (Aestek) [reply agent/consul/state/catalog.go]: My bad, these are strings

Comment (pierresouchay) [reply agent/consul/state/catalog.go]: Nope, this is a string

**Comment by linydquantil** [comment]
upup

**Comment by aashitvyas** [comment]
I am curious to know when this fix is getting merged into the master. We are doing upgrade on Consul from 1.0.6 to 1.3.1 and encountering issue mentioned in #4741. If this fix will get merge soon, we can have 1.3.1 on our environment rather then upgrading to 1.2.2

File: agent/consul/fsm/fsm.go [hunk 4]
```diff
@@ -59,22 +60,24 @@ type FSM struct {
...
 
-	gc *state.TombstoneGC
+	gc                 *state.TombstoneGC
+	nodeRenamingPolicy types.NodeRenamingPolicy
 }
 
 // New is used to construct a new FSM with a blank state.
-func New(gc *state.TombstoneGC, logOutput io.Writer) (*FSM, error) {
-	stateNew, err := state.NewStateStore(gc)
+func New(gc *state.TombstoneGC, logOutput io.Writer, nodeRenamingPolicy types.NodeRenamingPolicy) (*FSM, error) {
```
[the remark this answers is not in the document]
Comment (banks) [reply agent/consul/fsm/fsm.go]: Actually... forget the code sprawl, this indicates a much bigger issue - making FSM sate depend on outside input that is not guaranteed to be the same on all replicas. See my review comment.

**Comment by pierresouchay** [comment]
@banks I did think a while ago regarding this and thought about something: what about applying this on each agent individually ?

Thus, a node could specify if it has to be legacy, strict or "cloud ready"... Meaning the initial onode to be replaced would store the catalog the policy.

It make sense for instance in places where Consul in used in both bare-metal and cloudy environments. 

What do you think about that?

That's a bit problematic for wide policies (I mean, that does not depend upon a single node to take the decision), but for our case, I think it would work.

**Comment by banks** [comment]
I’ve just remembered we discussed this last week and came to a conclusion that I forgot to write down.

I’ll need to think about this more and expand upon it but it boils down to this:

 - Serf already allows nodes to be renamed safely - it has a mechanism to ensure a node rename is accepted by the cluster and will prevent nodes from claiming a name when another node disputes that it preset has that name. 
 - so in general if seed allows the new node to join then it’s pretty safe to allow it to register and take over that name
 - so we should forget having multiple modes and just make the normal behaviour be for the server to allow the rename of seed allows the join - exactly how we do that I’ve not looked into yet. 

Another reason to avoid multiple nodes is that external services (using ESM or otherwise) don’t have UUIDs so neither strict or dead node here would work for them and it’s not reasonably to call that “legacy”. We need to continue to support no UUID registrations for external nodes first class. 

Will come back to this when I get a chance, just wanted to note this while I remember!

**Comment by cwillbusch** [comment]
Any idea if this will fix the `Catalog.Register" RPC failed to server xxx.xxx.xxx.xxx:8300: rpc error making call: failed inserting node: Error while renaming Node ID: <node-id>: Node name xxx.xxx.xxx is reserved by node <node-id> with name xxx.xxx.xxx` error if you restore from a snapshot?

In our instance of consul if we restore from a snapshot whose node name is the same as the fresh consul node name we start seeing the error once we set an agent token. This is running consul 1.4 on Azure aks deployed using hashicorp's helm chart.

**Comment by banks** [comment]
@cwillbusch yeah that's a similar issue that would be fixed when we get this right. The complication there is that the snapshot restore is essentially time-travelling back to when a different node was registered in the same way!

If you manually deregister the node you should be able to get it to work in the interim. My proposed solution above (not this PR proposal) would fix it automatically since Serf layer would know that the old node is no longer around and so allow the re-registration.

**Comment by cwillbusch** [comment]
@banks Thanks for your clarification!

Could you let me know the proper way to manually deregister a node so that i can use this solution until your proposed solution is implemented?

**Comment by banks** [comment]
Heh that is actually a good question with a slightly sad answer @cwillbusch (CLI for this #2981 if interested).

You should be able to do so via the API though: https://www.consul.io/api/catalog.html#deregister-entity.

E.g.:

```
curl -i -X PUT -d '{"Node":"c1"}' 'http://localhost:8500/v1/catalog/deregister'
```

Note that despite the docs, this didn't seem to work when i attempted to use the Node's ID rather than its name - I think those docs predate Node IDs being a separate UUID. But this works I tested locally - the new node immediately is able to sync on the next attempt even without a restart.

**Comment by Nmishin** [comment]
Hi @banks, seems I have a related question here:
previously I have a Consul and clients 0.8.4 version, after upgrade Consul server to the 1.3.1 version I see in logs a lot of errors something like that:
` consul: RPC failed to server 10.10.9.5:8300: rpc error: rpc error making call: failed inserting node: Error while renaming Node ID: "e1d3e56d-9226-bf3a-571d-2a7e500c37dd": Node name dev-test-zcwf is reserved by node de084f26-6488-71a8-812f-e5f680ef4a17 with name dev-test-zcwf`
As I understand `dev-test-zcwf` node was recreated with the same name, but a new id, what should I do with this?
I didn't  restore Consul from a snapshot, only upgrade Consul with data saving. 
Do I need set `-disable-host-node-id` to false for preventing randomly generated node id?
Thanks.

**Comment by banks** [comment]
@Nmishin the answer is the same as above - assuming the old node actually doesn't exist, you either need to wait 72 hours or manually deregister the old node before you can add the new one with the curl command above.

Once we get the real fix in highlighted above, this problem should be able to solve itself automatically.

**Comment by kevstigneev** [comment]
> This is the first time that some state that is not governed by Raft is able to affect the decisions made inside the state machine! That sounds really dangerous!

Could not agree more.

Let's take it logically.

A node is uniquely identified in the cluster by its node_id and each node can have just one node_id. If that's true, one of the following must be true as well:

- several nodes can share the same node_name;
- node_id and node_name are effectively equal;
- a node can have several node_name's.

A node in Consul cluster is a data object. It is not the same as a Consul agent instance. Consul node and agent instance have disjoint lifecycles. node_id is meant to identify a data object, node name - an agent instance. If node_name is globally unique (space-time wise), the second option above (node_name ~ node_id) is fine.

However operators naturally want to use DNS names as node_name. DNS requires only spacial uniqueness. It does not formally require or guarantee that a name is not reused over time. It inhibits the second option.

From the remaining 2 options (id:name as N:1 and 1:N) the last one (multiple names per cluster) does not make sense. It means that a Consul agent can be represented by several Raft nodes.

It leaves just one option - several cluster nodes (data objects) can share the same node_name.

If a specific operation environment imposes extra restrictions, they should be addressed on top of the above. Not in the core.

**Comment by banks** [comment]
@kevstigneev thanks for the input. I think the overall decision here is clear so I'll close this.

For the record though:

> node_id is meant to identify a data object, node name - an agent instance.

This isn't quite right - node IDs are unique identifiers for the agent too - the reason we added them _as well_ as node names is that people like to rename their nodes from time to time so we wanted a separate stable ID. The issue is that it's impossible to make a truely stable ID - we try using hardware identifiers or failing that storing it on disk but then people find creative reasons that those change e.g. they rebuild an instance but keep same host name so now it has a new ID but same name.

That's the crux of the problem. Pierre's proposal here isn't bad per-se, it's just that we've realised we already have a mechanism (Serf) that can automatically resolve this conflict correctly without the complexity of explicit new behaviour options in the state machine or elsewhere.

We'll revisit this idea though if Serf can't solve for all cases as we expect.

**Pull request closed without merging**